cmd/kube-controller-manager/app/validatingadmissionpolicystatus.go

```diff
@@ -21,13 +21,26 @@ import (
 
 	pluginvalidatingadmissionpolicy "k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy"
 	"k8s.io/apiserver/pkg/cel/openapi/resolver"
+	genericfeatures "k8s.io/apiserver/pkg/features"
 	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/component-base/featuregate"
 	"k8s.io/controller-manager/controller"
+	"k8s.io/kubernetes/cmd/kube-controller-manager/names"
 	"k8s.io/kubernetes/pkg/controller/validatingadmissionpolicystatus"
 	"k8s.io/kubernetes/pkg/generated/openapi"
 )
 
-func startValidatingAdmissionPolicyStatusController(ctx context.Context, controllerContext ControllerContext) (controller.Interface, bool, error) {
+func newValidatingAdmissionPolicyStatusControllerDescriptor() *ControllerDescriptor {
+	return &ControllerDescriptor{
+		name:     names.ValidatingAdmissionPolicyStatusController,
+		initFunc: startValidatingAdmissionPolicyStatusController,
+		requiredFeatureGates: []featuregate.Feature{
```

Comment:
I don´t have the feeling this is a good idea, because right now it looks pretty neat and clean, but as soon as the code evolves and new features are added this is just yet another place to update the feature gates ... or what happen if the controller has only one part of the logic with a feature gate? we need to run the controller and we still need to add a feature gate inside the controller logic

some food for thoughts

Reply:
It depends on what tradeoff we want to make. If we want to have all the controller metadata and logic sprinkled everywhere or together. We already have feature gate logic sprinkled everywhere, so IMO this would not change that much. That is, one has to search for all occurrences of a feature gate to get the full picture.

 > or what happen if the controller has only one part of the logic with a feature gate? we need to run the controller and we still need to add a feature gate inside the controller logic

I think we would solve this the same way we do today. We can put a feature gate call inside the init function or inside the controller itself.

For the usual cases, we could reason about controllers without running them. That is, we could for example show which controllers are gated by which feature gates in kube-controller-manager's help

Reply:
the logic always going to be sprinkled everywhere, the feature gated code can be anything, a function or just part of a function ... 

> For the usual cases, we could reason about controllers without running them. That is, we could for example show which controllers are gated by which feature gates in kube-controller-manager's help

I think that the intent is good, but I'm afraid that the reality is different and it soon will get out of sync and we'll have an additional place interacting with the code that is not obvious for everybody

Reply:
I know it only helps a little but I have added a unit test to make sure people do not mix the old and the new approach.